README.rst
ExtractCode
===========

- license: Apache-2.0
- copyright: copyright (c) nexB. Inc. and others
- homepage_url: https://github.com/nexB/extractcode
- keywords: archive, extraction, libarchive, 7zip, scancode-toolkit, extractcode

Supports Windows, Linux and macOS on 64 bits processors and Python 3.6 to 3.9.


**ExtractCode is a (mostly) universal archive extractor.**

Install with::

    pip install extractcode[full]


Why another extractor?
----------------------

**it will extract!**

ExtractCode will extract things where other extractors may fail.

ExtractCode supports one of largest number of archive formats listed in the
long  `List of supported archive formats`_ found at the bottom of this document.

- Say you want to extract the tarball of the Linux kernel source code on Windows.
  It contains paths that are the same when ignoring the case and therefore will
  not extract OK on Windows: some file may be munged or the extract may file.

- Or a tarball (on any OS) may contain multiple times the exact same path. In
  these cases the paths showing up earlier in the archive may be "hidden" and
  overwritten by the same path showing up later in the archive giving the
  impression that there is only one file.

- Or an archive may be damaged a little but most files can still be extracted.

- Or the extracted files are such permissions that you cannot read them and are
  not owned by you.

- Or the archive may contain weird paths inluding relative paths that may be
  problematic to extract.

- Or the archive may contain special file types (character/device files) that
  may be problematic to extract.

- Or an archive may be a virtual disk or some file system(s) images that would
  typically need to be mounted to be accessed, and may require root access
  and guesswork to find out which partition and filesystem are at play and
  which driver to use.

In all these cases, ExtractCode will extract and try hard do the right thing to
obtain the actual archived content when other tools may fail.

It can also extract recursively any type of (nested) archives-in-archives.


As a downside, the extracted content may not be exactly what would be extracted
for a typical usage of the contained files: for instance some file may be
renamed, special files and symlinks are skipped, permissions and owners are
changed but this it is fine for prmary the use case which is analysis of file
content for software composition or forensic analysis.

Behind the scene, ExtractCode uses multiple tools such as:

- the Python standard library,
- a custom ctypes binding to libarchive,
- the 7zip command line tool, and
- optionally libguestfs on Linux.

With these, it is possible to extract a large number of common and less common
archives and compressed file types. ExtractCode tries to extract things in the
same way on all supported OSes, including auto-renaming files that would have
invalid, non-extractible names on certain filesystems or when there are multiple
copies of the same path in a given archive (which is possible in a tar).

The extraction is driven from  a "voting" system that considers the file
extension(s) and name, the filetype and mimetype (using a ctypes binding to
libmagic) to select the most appropriate extractor or decompressor function.
It can handle multi-level archives such as tar.gz and can extract recursively
any nested archives.

Visit https://aboutcode.org and https://github.com/nexB/ for support and download.


We run CI tests on:

 - Azure pipelines https://dev.azure.com/nexB/extractcode/_build


Installation
------------

To install this package with its full capability (where the binaries for
7zip and libarchive are installed), use the `full` extra option::

    pip install extractcode[full]

If you want to use the version of binaries (possibly) provided by your operating
system, use the `minimal` option::

    pip install extractcode

In this case, you will need to provide a working and compatible libarchive and
7zip installed and configured in one of these ways such that ExtractCode can
find them:

- **a typecode-libarchive and typecode-7z plugin**: See the standard ones at
  https://github.com/nexB/scancode-plugins/tree/main/builtins
  These can either bundle a libarchive library, a 7z executable or expose a
  system-installed libraries.
  It does so by providing plugin entry points as ``scancode_location_provider``
  for ``extractcode_libarchive`` that should point to a ``LocationProviderPlugin``
  subclass with a ``get_locations()`` method that must return a mapping with
  this key:

    - 'extractcode.libarchive.dll': the absolute path to a **libarchive** shared object/DLL

  See for example:

    - https://github.com/nexB/scancode-plugins/blob/4da5fe8a5ab1c87b9b4af9e54d7ad60e289747f5/builtins/extractcode_libarchive-linux/setup.py#L40
    - https://github.com/nexB/scancode-plugins/blob/4da5fe8a5ab1c87b9b4af9e54d7ad60e289747f5/builtins/extractcode_libarchive-linux/src/extractcode_libarchive/__init__.py#L17

  And in the same way, the ``scancode_location_provider`` for ``extractcode_7zip``
  should point to a ``LocationProviderPlugin`` subclass with a ``get_locations()``
  method that must return a mapping with this key:

    - 'extractcode.sevenzip.exe': the absolute path to a **7zip** executable

  See for example:

    - https://github.com/nexB/scancode-plugins/blob/4da5fe8a5ab1c87b9b4af9e54d7ad60e289747f5/builtins/extractcode_7z-linux/setup.py#L40
    - https://github.com/nexB/scancode-plugins/blob/4da5fe8a5ab1c87b9b4af9e54d7ad60e289747f5/builtins/extractcode_7z-linux/src/extractcode_7z/__init__.py#L18

- use **environment variables** to point to installed binaries:

    - EXTRACTCODE_LIBARCHIVE_PATH: the absolute path to a libarchive DLL
    - EXTRACTCODE_7Z_PATH: the absolute path to a 7zip executable


- **a system-installed libarchive and 7zip executable** available in the system **PATH**.


The supported binary tools versions are:

- libarchive  3.5.x
- 7zip 16.5.x

Development
-----------

To set up the development environment::

    ./configure --dev
    source venv/bin/activate


To run unit tests::

    pytest -vvs -n 2


To clean up development environment::

    ./configure --clean


To run the command line tool in the activated environment::

    ./extractcode -h


Configuration with environment variables
----------------------------------------

ExtractCode will use these environment variables if set:

- EXTRACTCODE_LIBARCHIVE_PATH : the path to the ``libarchive.so`` libarchive
  shared library used to support some of the archive formats. If not provided,
  ExtractCode will look for a plugin-provided libarchive library path. See
  https://github.com/nexB/scancode-plugins/tree/main/builtins for such plugins.
  If no plugin contributes libarchive, then a final attempt is made to look for
  it in the PATH using standard DLL loading techniques.

- EXTRACTCODE_7Z_PATH : the path to the ``7z`` 7zip executable used to support
  some of the archive formats. If not provided, ExtractCode will look for a
  plugin-provided 7z executable path. See
  https://github.com/nexB/scancode-plugins/tree/main/builtins for such plugins.
  If no plugin contributes 7z, then a final attempt is made to look for
  it in the PATH.

- EXTRACTCODE_GUESTFISH_PATH : the path to the ``guestfish`` tool from
  libguestfs to use to extract VM images. If not provided, ExtractCode will look
  in the PATH for an installed ``guestfish`` executable instead.



Adding support for VM images extraction
---------------------------------------

Adding support for VM images requires the manual installation of the
libguestfs-tools system package. This is suported only on Linux.
On Debian and Ubuntu you can use this command::

    sudo apt-get install libguestfs-tools


On Ubuntu only, an additional manual step is required as the kernel executable
file cannot be read by users as required by libguestfish.

Run this command as a temporary and immediate fix::

    sudo chmod 0644 /boot/vmlinuz-*
    for k in /boot/vmlinuz-*
        do sudo dpkg-statoverride --add --update root root 0644 /boot/vmlinuz-$k
    done

You likely want both this temporary fix and a more permanent fix; otherwise each
kernel update will revert to the default permissions and ExtractCode will stop
working for VM images extraction.

Therefore follow these instructions:

1. As sudo, create the file /etc/kernel/postinst.d/statoverride with this
content, devised by Kees Cook (@kees) in
https://bugs.launchpad.net/ubuntu/+source/linux/+bug/759725/comments/3 ::

    #!/bin/sh
    version="$1"
    # passing the kernel version is required
    [ -z "${version}" ] && exit 0
    dpkg-statoverride --update --add root root 0644 /boot/vmlinuz-${version}

2. Set executable permissions::

    sudo chmod +x /etc/kernel/postinst.d/statoverride

See also these links for a complete discussion:

    - https://bugs.launchpad.net/ubuntu/+source/linux/+bug/759725
    - https://bugzilla.redhat.com/show_bug.cgi?id=1670790
    - https://bugs.launchpad.net/ubuntu/+source/libguestfs/+bug/1813662/comments/24


Alternative
-----------

These other tools are related and were considered before creating ExtractCode:

These tools provide built-in, original extraction capabilities:

- https://libarchive.org/ (integrated in ExtractCode) (BSD license)
- https://www.7-zip.org/ (integrated in ExtractCode) (LGPL license)
- https://theunarchiver.com/command-line (maintenance status unknown) (LGPL license)

These tools are command line tools  wrapping other extraction tools and are
similar to ExtractCode but with different goals:

- https://github.com/wummel/patool (wrapper on many CLI tools) (GPL license)
- https://github.com/dtrx-py/dtrx (wrapper on a few CLI tools) (recently revived) (GPL license)



List of supported archive formats
-------------------------------------

ExtractCode can extract the folowing archives formats:

Archive format kind: docs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  name: Office doc
     - extensions: .docx, .dotx, .docm, .xlsx, .xltx, .xlsm, .xltm, .pptx, .ppsx, .potx, .pptm, .potm, .ppsm, .odt, .odf, .sxw, .stw, .ods, .ots, .sxc, .stc, .odp, .otp, .odg, .otg, .sxi, .sti, .sxd, .sxg, .std, .sdc, .sda, .sdd, .smf, .sdw, .sxm, .stw, .oxt, .sldx, .epub
     - filetypes : zip archive, microsoft word 2007+, microsoft excel 2007+, microsoft powerpoint 2007+
     - mimetypes : application/zip, application/vnd.openxmlformats

  name: Dia diagram doc
     - extensions: .dia
     - filetypes : gzip compressed
     - mimetypes : application/gzip

  name: Graffle diagram doc
     - extensions: .graffle
     - filetypes : gzip compressed
     - mimetypes : application/gzip

  name: SVG Compressed doc
     - extensions: .svgz
     - filetypes : gzip compressed
     - mimetypes : application/gzip

Archive format kind: regular
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  name: Tar
     - extensions: .tar
     - filetypes : .tar, tar archive
     - mimetypes : application/x-tar

  name: Zip
     - extensions: .zip, .zipx
     - filetypes : zip archive
     - mimetypes : application/zip

  name: Java archive
     - extensions: .war, .sar, .ear
     - filetypes : zip archive
     - mimetypes : application/zip, application/java-archive

  name: xz
     - extensions: .xz
     - filetypes : xz compressed
     - mimetypes : application/x-xz

  name: lzma
     - extensions: .lzma
     - filetypes : lzma compressed
     - mimetypes : application/x-xz

  name: Gzip
     - extensions: .gz, .gzip, .wmz, .arz
     - filetypes : gzip compressed, gzip compressed data
     - mimetypes : application/gzip

  name: bzip2
     - extensions: .bz, .bz2, bzip2
     - filetypes : bzip2 compressed
     - mimetypes : application/x-bzip2

  name: lzip
     - extensions: .lzip
     - filetypes : lzip compressed
     - mimetypes : application/x-lzip

  name: RAR
     - extensions: .rar
     - filetypes : rar archive
     - mimetypes : application/x-rar

  name: ar archive
     - extensions: .ar
     - filetypes : current ar archive
     - mimetypes : application/x-archive

  name: 7zip
     - extensions: .7z
     - filetypes : 7-zip archive
     - mimetypes : application/x-7z-compressed

  name: cpio
     - extensions: .cpio
     - filetypes : cpio archive
     - mimetypes : application/x-cpio

  name: Z
     - extensions: .z
     - filetypes : compress'd data
     - mimetypes : application/x-compress

Archive format kind: regular_nested
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  name: Tar xz
     - extensions: .tar.xz, .txz, .tarxz
     - filetypes : xz compressed
     - mimetypes : application/x-xz

  name: Tar lzma
     - extensions: tar.lzma, .tlz, .tarlz, .tarlzma
     - filetypes : lzma compressed
     - mimetypes : application/x-lzma

  name: Tar gzip
     - extensions: .tgz, .tar.gz, .tar.gzip, .targz, .targzip, .tgzip
     - filetypes : gzip compressed
     - mimetypes : application/gzip

  name: Tar lzip
     - extensions: .tar.lz, .tar.lzip
     - filetypes : lzip compressed
     - mimetypes : application/x-lzip

  name: Tar lz4
     - extensions: .tar.lz4
     - filetypes : lz4 compressed
     - mimetypes : application/x-lz4

  name: Tar zstd
     - extensions: .tar.zst, .tar.zstd
     - filetypes : zstandard compressed
     - mimetypes : application/x-zstd

  name: Tar bzip2
     - extensions: .tar.bz2, .tar.bz, .tar.bzip, .tar.bzip2, .tbz, .tbz2, .tb2, .tarbz2
     - filetypes : bzip2 compressed
     - mimetypes : application/x-bzip2

  name: lz4
     - extensions: .lz4
     - filetypes : lz4 compressed
     - mimetypes : application/x-lz4

  name: zstd
     - extensions: .zst, .zstd
     - filetypes : zstandard compressed
     - mimetypes : application/x-zstd

  name: Tar 7zip
     - extensions: .tar.7z, .tar.7zip, .t7z
     - filetypes : 7-zip archive
     - mimetypes : application/x-7z-compressed

  name: Tar Z
     - extensions: .tz, .tar.z, .tarz
     - filetypes : compress'd data
     - mimetypes : application/x-compress


Archive format kind: package
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  name: Ruby Gem package
     - extensions: .gem
     - filetypes : .tar, tar archive
     - mimetypes : application/x-tar

  name: Android app
     - extensions: .apk
     - filetypes : zip archive
     - mimetypes : application/zip

  name: Android library
     - extensions: .aar
     - filetypes : zip archive
     - mimetypes : application/zip

  name: Mozilla extension
     - extensions: .xpi
     - filetypes : zip archive
     - mimetypes : application/zip

  name: iOS app
     - extensions: .ipa
     - filetypes : zip archive
     - mimetypes : application/zip

  name: Springboot Java Jar package
     - extensions: .jar
     - filetypes : bourne-again shell script executable (binary data)
     - mimetypes : text/x-shellscript

  name: Java Jar package
     - extensions: .jar, .zip
     - filetypes : java archive
     - mimetypes : application/java-archive

  name: Java Jar package
     - extensions: .jar
     - filetypes : zip archive
     - mimetypes : application/zip

  name: Python package
     - extensions: .egg, .whl, .pyz, .pex
     - filetypes : zip archive
     - mimetypes : application/zip

  name: Microsoft cab
     - extensions: .cab
     - filetypes : microsoft cabinet
     - mimetypes : application/vnd.ms-cab-compressed

  name: Microsoft MSI Installer
     - extensions: .msi
     - filetypes : msi installer
     - mimetypes : application/x-msi

  name: Apple pkg or mpkg package installer
     - extensions: .pkg, .mpkg
     - filetypes : xar archive
     - mimetypes : application/octet-stream

  name: Xar archive v1
     - extensions: .xar
     - filetypes : xar archive
     - mimetypes : application/octet-stream, application/x-xar

  name: Nuget
     - extensions: .nupkg
     - filetypes : zip archive, microsoft ooxml
     - mimetypes : application/zip, application/octet-stream

  name: Static Library
     - extensions: .a, .lib, .out, .ka
     - filetypes : current ar archive, current ar archive random library
     - mimetypes : application/x-archive

  name: Debian package
     - extensions: .deb, .udeb
     - filetypes : debian binary package
     - mimetypes : application/vnd.debian.binary-package, application/x-archive

  name: RPM package
     - extensions: .rpm, .srpm, .mvl, .vip
     - filetypes : rpm 
     - mimetypes : application/x-rpm

  name: Apple dmg
     - extensions: .dmg, .sparseimage
     - filetypes : zlib compressed
     - mimetypes : application/zlib

Archive format kind: file_system
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  name: ISO CD image
     - extensions: .iso, .udf, .img
     - filetypes : iso 9660 cd-rom, high sierra cd-rom
     - mimetypes : application/x-iso9660-image

  name: SquashFS disk image
     - extensions: 
     - filetypes : squashfs
     - mimetypes : 

  name: QEMU QCOW2 disk image
     - extensions: .qcow2, .qcow, .qcow2c, .img
     - filetypes : qemu qcow2 image, qemu qcow image
     - mimetypes : application/octet-stream

  name: VMDK disk image
     - extensions: .vmdk
     - filetypes : vmware4 disk image
     - mimetypes : application/octet-stream

  name: VirtualBox disk image
     - extensions: .vdi
     - filetypes : virtualbox disk image
     - mimetypes : application/octet-stream

Archive format kind: patches
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  name: Patch
     - extensions: .diff, .patch
     - filetypes : diff, patch
     - mimetypes : text/x-diff

Archive format kind: special_package
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  name: InstallShield Installer
     - extensions: .exe
     - filetypes : installshield
     - mimetypes : application/x-dosexec

  name: Nullsoft Installer
     - extensions: .exe
     - filetypes : nullsoft installer
     - mimetypes : application/x-dosexec




CHANGELOG.rst
Changelog
=========

v31.0.0
--------

- Do not install patch.py by default. Instead this is now an extra


v30.0.0
--------

- Update to the latest skeleton. The virtualenv is now created under the venv
  directory with the ./configure --dev
- Switch back to semver versioning. Calver was not really needed here.
- Do not crash if there is a corrupted archive with the --replace-originals
  option

Thank you to:

- Jono Yang @JonoYang
- Bryan Sutula @sutula
- Smascer @Smascer
- Chin-Yeung Li @chinyeungli

v21.7.23
--------

- Spaces in file and directory names are no longer replaced by underscores when
  files and directories are extracted from archives.

v21.6.2
-------

- Add new --list-formats command line option to list supported archive formats
- Add new exttractcode.api.extract_archive() API function to extract a single
  archive file of any supported format, non recursively.


v21.6.1
-------

- Add support for VMDK, QCOW and VDI VM image filesystems extraction
- Add new configuration mechanism to get third-party binary paths:

  - Use an environment variable
  - Or use a plugin-provided path
  - Or use well-known system installation locations
  - Or use the system PATH
  - Or fail with an informative error message

- Update to use latest skeleton


v21-2-24
----------

- Fix incorrect documentation link


v21-1-21
----------

- Fix bug related to CommonCode libraries loading
- Improve the extra requirements
- Set minimum version for dependencies
- Improve documentation
- Reorganize tests files


v21-1-15
----------

- Drop support for Python 2
- Use the latest CommonCode and TypeCode libraries
- Add azure-pipelines CI support


v20.10
------

- Initial release as a split from ScanCode toolkit


src/extractcode/vmimage.py
#
# Copyright (c) nexB Inc. and others. All rights reserved.
# ScanCode is a trademark of nexB Inc.
# SPDX-License-Identifier: Apache-2.0
# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
# See https://github.com/nexB/extractcode for support or download.
# See https://aboutcode.org for more information about nexB OSS projects.
#

import logging
import os
import pathlib
import shutil
import warnings

import attr

from commoncode import fileutils
from commoncode.system import on_linux
from commoncode.text import as_unicode

from extractcode import ExtractErrorFailedToExtract

"""
Support to extract Virtual Machine image formats and the filesystem(s) they
contain. This is based on libguestfs-tools and is tested only on Linux. Works
only if libguestfs tool guestfish:
- has its path in the "EXTRACTCODE_GUESTFISH_PATH" environment variable.
- or is in the system PATH.

See https://libguestfs.org/
"""

logger = logging.getLogger(__name__)

TRACE = False

if TRACE:
    import sys
    logging.basicConfig(stream=sys.stdout)
    logger.setLevel(logging.DEBUG)

GUESTFISH_NOT_FOUND = (
    'WARNING: guestfish executable is not installed. '
    'Unable to extract virtual machine image: you need to install the '
    'guestfish tool from libguestfs and extra FS drivers as needed. '
    'See the ExtractCode README.rst at '
    'https://github.com/nexB/extractcode/blob/main/README.rst '
    'and https://libguestfs.org/ for details.'
)

GUESTFISH_KERNEL_NOT_READABLE = (
'''libguestfs requires the kernel executable to be readable.
This is the case by default on most Linux distributions except on Ubuntu.
Please follow the ExtractCode installation instructions in the README.rst at:
https://github.com/nexB/extractcode/blob/main/README.rst '
''')

EXTRACTCODE_GUESTFISH_PATH_ENVVAR = 'EXTRACTCODE_GUESTFISH_PATH'


def get_command(env_var=EXTRACTCODE_GUESTFISH_PATH_ENVVAR, command='guestfish'):
    """
    Return the location to the guestfish command or None.
    """
    cmd_loc = os.environ.get(env_var, None)
    if cmd_loc and os.path.exists(cmd_loc):
        return cmd_loc

    cmd_loc = shutil.which(command) or None
    if not cmd_loc:
        warnings.warn(GUESTFISH_NOT_FOUND)

    return cmd_loc


def check_linux_kernel_is_readable():
    """
    Return True if the kernel executable file can be read. This is required by
    guestfish and libguestfs and this is an oddity mostly on Ubuntu.

    See:
      - https://bugs.launchpad.net/ubuntu/+source/linux/+bug/759725
      - https://bugzilla.redhat.com/show_bug.cgi?id=1670790
      - https://bugs.launchpad.net/ubuntu/+source/libguestfs/+bug/1813662
    """

    if on_linux:
        kernels = list(pathlib.Path('/boot').glob('vmlinuz-*'))
        if not kernels:
            raise ExtractErrorFailedToExtract(GUESTFISH_KERNEL_NOT_READABLE)
        for kern in kernels:
            if not os.access(kern, os.R_OK):
                raise ExtractErrorFailedToExtract(
                    f'Unable to read kernel at: {kern}.\n'
                    f'{GUESTFISH_KERNEL_NOT_READABLE}'
                )


@attr.s
class VmImage:
    location = attr.ib()
    image_format = attr.ib()
    guestfish_command = attr.ib()

    @classmethod
    def from_file(cls, location):
        """
        Build a new VMImage from the file at location.
        Raise excptions on errors.
        """
        if not on_linux:
            raise ExtractErrorFailedToExtract(
                'VM Image extraction only supported on Linux.')

        # check_linux_kernel_is_readable()

        assert location
        abs_location = os.path.abspath(os.path.expanduser(location))

        if not os.path.exists(abs_location):
            raise ExtractErrorFailedToExtract(
                f'The system cannot find the path specified: {abs_location}')

        supported_gfs_formats_by_extension = {
            '.qcow2': 'qcow2',
            '.qcow2c': 'qcow2',
            '.qcow': 'qcow2',
            '.img': 'qcow2',
            '.vmdk': 'vmdk',
            '.vdi': 'vdi',
        }

        extension = fileutils.file_extension(location)
        image_format = supported_gfs_formats_by_extension.get(extension)

        if not image_format:
            raise ExtractErrorFailedToExtract(
                f'Unsupported VM image format: {location}')

        cmd_loc = get_command()
        if not cmd_loc:
            raise ExtractErrorFailedToExtract(GUESTFISH_NOT_FOUND)

        return cls(
            location=location,
            image_format=image_format,
            guestfish_command=cmd_loc,
        )

    def listfs(self, skip_partitions=('swap',)):
        """
        Return a list of (filesystem /partition/ device path, filesystem type)
        for each filesystem found in this image.

        Skip the partitions names list ``skip_partitions``.

        We run guestfish for this:
            $ guestfish --ro add foo.qcow2 : run : list-filesystems
            /partition/sda1: ext4
        """
        args = [
            '--ro',
            f'--format={self.image_format}',
            '--add' , self.location,
            'run',
            ':', 'list-filesystems',
        ]
        stdout = self.run_guestfish(args)

        filesystems = []
        entries = stdout.strip().splitlines(False)
        for entry in entries:
            entry = entry.strip()
            if not entry:
                continue
            if ':' in entry:
                partition, _, fstype = entry.partition(':')
                fstype = fstype.strip()
            else:
                partition = entry
                fstype = None

            if any(s in partition for s in skip_partitions):
                continue

            filesystems.append((partition, fstype,))

        return filesystems

    def extract_image(self, target_tarball):
        """
        Extract all files from this VM image to the ``target_tarball`` file as a
        gzipped-compressed tarball (.tar.gz). Raise Exception on errors.
        """
        args = [
            '--ro',
            '--inspector',
            f'--format={self.image_format}',
            '--add', self.location,
            'tar-out', '/', target_tarball, 'compress:gzip',
        ]

        self.run_guestfish(args)

    def extract_partition(self, partition, target_tarball):
        """
        Extract all files from a single ``partition`` of this VM image to the
        ``target_tarball`` file as a gzipped-compressed tarball (.tar.gz). Raise
        exception on errors.
        """
        # TODO: there could be devices/partitions we do not want to extract?
        # guestfish --ro add foo.qcow2 : run : mount /dev/sda1 / : tar-out /etc foo.tgz compress:gzip

        args = [
            '--ro',
            f'--format={self.image_format}',
            '--add', self.location,
            'run',
            ':', 'mount', partition, '/',
            ':', 'tar-out', '/', target_tarball, 'compress:gzip',
        ]
        self.run_guestfish(args)

    def run_guestfish(self, args, timeout=None):
        """
        Run guestfish with ``args`` arguments and am optional ``timeout`` in
        seconds. Return stdout as a unicode string. Raise Exception on error.
        """
        import subprocess
        full_args = [self.guestfish_command] + args
        try:
            stdout = subprocess.check_output(
                full_args,
                timeout=timeout,
                stderr=subprocess.STDOUT,
            )
        except subprocess.CalledProcessError as cpe:
            args = ' '.join([self.guestfish_command] + args)
            output = as_unicode(cpe.output)
            error = (
                f'Failed to run guestfish to extract VM image: {args}\n'
                f'output: {output}'
            )
            raise ExtractErrorFailedToExtract(error)

        return as_unicode(stdout)


def extract(location, target_dir, as_tarballs=False, skip_symlinks=True):
    """
    Extract all files from a guestfish-supported VM image archive file at
    ``location`` in the ``target_dir`` directory.
    Return a list of warning messages if any or an empty list.
    Raise Exception on errors.

    Optionally only extract the intermediate tarballs if ``as_tarball`` is True.
    Otherwise, extract to intermediate tarballs and then extract each tarballs
    to the final directory.

    Optionally skip extracting symlinks with ``skip_symlinks``.

    This works only on Linux.
    """
    assert target_dir
    abs_target_dir = os.path.abspath(os.path.expanduser(target_dir))
    if not os.path.exists(abs_target_dir) or not os.path.isdir(abs_target_dir):
        raise ExtractErrorFailedToExtract(
            f'The system cannot find the target directory path '
            f'specified: {target_dir}')

    vmimage = VmImage.from_file(location)

    warnings = []

    filename = fileutils.file_name(vmimage.location)

    # try a plain extract first
    try:

        if not as_tarballs:
            intermediate_dir = fileutils.get_temp_dir(
                prefix='extractcode-vmimage')
            tdir = intermediate_dir
        else:
            tdir = target_dir

        target_tarball = os.path.join(tdir, f'{filename}.tar.gz')
        vmimage.extract_image(target_tarball=target_tarball)

        if not as_tarballs:
            # extract the temp tarball to the final location
            warns = extract_image_tarball(
                tarball=target_tarball,
                target_dir=target_dir,
                skip_symlinks=skip_symlinks,
            )
            warnings.extend(warns)

    except ExtractErrorFailedToExtract as e:
        print('Cannot extract VM Image filesystems as a single file tree.')

        warnings.append(
            f'Cannot extract VM Image filesystems as a single file tree:\n{e}')
        # fall back to file system extraction, one partition at a time
        partitions = vmimage.listfs()
        if not partitions:
            raise

        if len(partitions) == 1:
            # we can safely extract this to a root / dir as we have only one partition
            partition, _parttype = partitions[0]
            if not as_tarballs:
                tdir = fileutils.get_temp_dir(prefix='extractcode-vmimage')
            else:
                tdir = target_dir

            target_tarball = os.path.join(tdir, f'{filename}.tar.gz')
            vmimage.extract_partition(
                partition=partition,
                target_tarball=target_tarball,
            )

            if not as_tarballs:
                # extract the temp tarball to the final location
                warns = extract_image_tarball(
                    tarball=target_tarball,
                    target_dir=target_dir,
                    skip_symlinks=skip_symlinks,
                )
                warnings.extend(warns)
        else:
            # with multiple partitions, we extract each partition to a unique
            # base name based after the partition device name

            for partition, _parttype in partitions:
                base_name = partition.replace('/', '-')

                if not as_tarballs:
                    tdir = fileutils.get_temp_dir(prefix='extractcode-vmimage')
                else:
                    tdir = target_dir

                partition_tarball = os.path.join(
                    tdir,
                    f'{filename}-{base_name}.tar.gz',
                )
                vmimage.extract_partition(
                    partition=partition,
                    target_tarball=partition_tarball,
                )

                if not as_tarballs:
                    # extract the temp tarball to the final location
                    # which is a new subdirectory
                    partition_target_dir = os.path.join(target_dir, base_name)
                    fileutils.create_dir(partition_target_dir)
                    warns = extract_image_tarball(
                        tarball=target_tarball,
                        target_dir=partition_target_dir,
                        skip_symlinks=skip_symlinks,
                    )
                    warnings.extend(warns)

    return warnings


def extract_image_tarball(tarball, target_dir, skip_symlinks=True):
    """
    Extract an intermediate image ``tarball`` to its final ``target_dir``
    directory. Optionally skip extracting symlinks with ``skip_symlinks``.
    Return a list of warning messages. Raise Exception on errors.
    """
    from extractcode.libarchive2 import extract
    return extract(
        location=tarball,
        target_dir=target_dir,
        skip_symlinks=skip_symlinks,
    )
